### PR TITLE
iAL parallel

### PR DESCRIPTION
**Pull request by yichen88**

### CODE REVIEW

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.action.simulator.parallel;
+
+import com.powsybl.commons.PowsyblException;
+
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+public class Filtration {
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java]: I'm not really sure Filtration is the correct word here.

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    private static final Pattern PATTERN = Pattern.compile("\\d+/\\d+");
+
+    private final Integer x;
+    private final Integer y;
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java]: Is Integer required here or primitive type is also ok

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        boolean valid = PATTERN.matcher(filtration).find();
+        if (!valid) {
+            throw new PowsyblException(filtration + " is not valid");
+        }
+
+        String[] split = filtration.split("/");
+        x = Integer.valueOf(split[0]);
+        y = Integer.valueOf(split[1]);
+
+        if (x > y || x < 1 || y < 1) {
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java]: Is is OK if x = y?

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+import static com.powsybl.action.simulator.tools.ActionSimulatorToolConstants.*;
+
+public class ParallelLoadFlowActionSimulator extends LoadFlowActionSimulator {
+
+    private static final String ITOOLS_PRG = "itools";
+
+    private final int para;
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: What "para" means? Could you find a better name?

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    private final CommandLine commandLine;
+
+    private static final String SUB_TASK_ID = "sas"; // sub_action_simulator
+
+    public ParallelLoadFlowActionSimulator(Network network, ToolRunningContext context, int para, CommandLine commandLine) {
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Same remark here

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        this(network, context, para, commandLine, LoadFlowActionSimulatorConfig.load(), false, Collections.emptyList());
+    }
+
+    public ParallelLoadFlowActionSimulator(Network network, ToolRunningContext context, int para, CommandLine commandLine,
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Same remark here

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                                           LoadFlowActionSimulatorConfig config, boolean applyIfSolved, LoadFlowActionSimulatorObserver... observers) {
+        this(network, context, para, commandLine, config, applyIfSolved, Arrays.asList(observers));
+    }
+
+    public ParallelLoadFlowActionSimulator(Network network, ToolRunningContext context, int para, CommandLine commandLine,
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Same remark here

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                                           LoadFlowActionSimulatorConfig config, boolean applyIfSolved, List<LoadFlowActionSimulatorObserver> observers) {
+        super(network, context.getLongTimeExecutionComputationManager(), config, applyIfSolved, observers);
+        this.para = para;
+        this.commandLine = Objects.requireNonNull(commandLine);
+    }
+
+    @Override
+    public String getName() {
+        return "parallel loadflow actionsimulator";
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Add a space or a dash between action and simulator

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+    public ParallelLoadFlowActionSimulator(Network network, ToolRunningContext context, int para, CommandLine commandLine,
+                                           LoadFlowActionSimulatorConfig config, boolean applyIfSolved, LoadFlowActionSimulatorObserver... observers) {
+        this(network, context, para, commandLine, config, applyIfSolved, Arrays.asList(observers));
+    }
+
+    public ParallelLoadFlowActionSimulator(Network network, ToolRunningContext context, int para, CommandLine commandLine,
+                                           LoadFlowActionSimulatorConfig config, boolean applyIfSolved, List<LoadFlowActionSimulatorObserver> observers) {
+        super(network, context.getLongTimeExecutionComputationManager(), config, applyIfSolved, observers);
+        this.para = para;
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: You probably have to add a check on this parameter

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    @Override
+    public String getName() {
+        return "parallel loadflow actionsimulator";
+    }
+
+    @Override
+    public void start(ActionDb actionDb, List<String> contingencyIds) {
+        ComputationManager manager = super.getComputationManager();
+        ExecutionEnvironment itoolsEnvironment = new ExecutionEnvironment(Collections.emptyMap(), "subTask_", config.isDebug());
+        int realPara = contingencyIds.size() > this.para ? this.para : contingencyIds.size();
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Seems "para" is contingencyCount
Rename also realPara to increase readability

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+import java.util.concurrent.CompletableFuture;
+
+import static com.powsybl.action.simulator.tools.ActionSimulatorToolConstants.*;
+
+public class ParallelLoadFlowActionSimulator extends LoadFlowActionSimulator {
+
+    private static final String ITOOLS_PRG = "itools";
+
+    private final int para;
+    private final CommandLine commandLine;
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: You probably should not depend on the command line here

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+    @Override
+    public void start(ActionDb actionDb, String... contingencyIds) {
+        start(actionDb, Arrays.asList(contingencyIds));
+    }
+
+    private class SubTaskHandler extends AbstractExecutionHandler<SecurityAnalysisResult> {
+
+        private final Integer index;
+        private final Integer total;
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Not really sure that null is allowed here

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+        SubTaskHandler(Integer x, Integer y) {
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: Here neither

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/tools/ActionSimulatorTool.java [hunk 14]
```diff
@@ -225,8 +230,18 @@ public void run(CommandLine line, ToolRunningContext context) throws Exception {
...
 
             // action simulator
-            ActionSimulator actionSimulator = new LoadFlowActionSimulator(network, context.getShortTimeExecutionComputationManager(),
-                    config, line.hasOption(APPLY_IF_SOLVED_VIOLATIONS), observers);
+            ActionSimulator actionSimulator = null;
+            if (line.hasOption(PARALLEL)) {
+                checkOptionsInParallel(line);
+                int para = Integer.parseInt(line.getOptionValue(PARALLEL));
+                actionSimulator = new ParallelLoadFlowActionSimulator(network, context, para, line, config, applyIfSolved, observers);
+            } else if (line.hasOption(SUB_CONTINGENCIES)) {
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/tools/ActionSimulatorTool.java]: Not really sure that the name of this option is really good

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/tools/ActionSimulatorTool.java [hunk 15]
```diff
@@ -225,8 +230,18 @@ public void run(CommandLine line, ToolRunningContext context) throws Exception {
             }
 
             // action simulator
-            ActionSimulator actionSimulator = new LoadFlowActionSimulator(network, context.getShortTimeExecutionComputationManager(),
-                    config, line.hasOption(APPLY_IF_SOLVED_VIOLATIONS), observers);
+            ActionSimulator actionSimulator = null;
+            if (line.hasOption(PARALLEL)) {
```
Comment (mathbagu) [inline action/action-simulator/src/main/java/com/powsybl/action/simulator/tools/ActionSimulatorTool.java]: does it conflict with the parallel option of itools

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisResultMerger.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.security;
+
+import java.util.Arrays;
+import java.util.Collections;
+
+public final class SecurityAnalysisResultMerger {
+
+    public static SecurityAnalysisResult merge(SecurityAnalysisResult[] results) {
```
Comment (mathbagu) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisResultMerger.java]: Add Objects.requireNonNull(results)

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java (hunk 3 above)
Comment (yichen88) [reply action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/Filtration.java]: yes. and tested in tests. ex: 3/3 means the last part.

File: action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java (hunk 4 above)
Comment (yichen88) [reply action/action-simulator/src/main/java/com/powsybl/action/simulator/parallel/ParallelLoadFlowActionSimulator.java]: The number of sub-processes expected to launch simulator in parallel mode.
realPara is the used when total contingencies is smaller than the expected parallel counts,

**Pull request merged**